core/git_command.py

```diff
@@ -110,6 +110,74 @@ def communicate(self, stdin):
         return self.stdout, self.stderr
 
 
+if False:
+    from typing import Dict, Iterator, Optional
+
+
+def repo_path_for_view_if_cached(view):
+    # type: (sublime.View) -> Optional[str]
+    repo_path = view.settings().get("git_savvy.repo_path")
+    if repo_path and os.path.exists(repo_path):
+        return repo_path
+
+    for path in search_paths(view):
+        try:
+            return repo_paths[path]
+        except KeyError:
+            ...
+
+    return None
+
+
+def repo_path_for_view(view):
+    # type: (sublime.View) -> Optional[str]
+    repo_path = view.settings().get("git_savvy.repo_path")
+    if repo_path and os.path.exists(repo_path):
+        return repo_path
+
+    for path in search_paths(view):
+        repo_path = map_path_to_repo_path(path)
+        if repo_path:
+            return repo_path
+
+    return None
+
+
+repo_paths = {}  # type: Dict[str, str]
+
+
+def map_path_to_repo_path(path):
+    # type: (str) -> Optional[str]
+    try:
+        return repo_paths[path]
+    except KeyError:
+        repo_path = call_git_toplevel(path)
+        if repo_path:
+            repo_path = os.path.normpath(repo_path)
+            repo_paths[path] = repo_path
+        return repo_path
+
+
+def call_git_toplevel(folder):
+    # type: (str) -> Optional[str]
+    return GitCommand().git(
+        'rev-parse', '--show-toplevel', working_dir=folder, throw_on_stderr=False
+    ).strip() or None
+
+
+def search_paths(view):
```

Comment:
`search_paths` is probably not reusable code. It also currently introduces the miserable edge-case behavior for untracked projects, in that we first call git for the filepath, then for the project folder. But since it is an untracked project bot calls don't lead to a value, and we can't cache `None`s. So we do two toplevel calls over and over. (To the effect that we have a yanky experience for the untracked projects, basically when we ideally should consume no resources at all.)